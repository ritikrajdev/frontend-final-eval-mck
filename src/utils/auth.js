export function getAuthToken() {
  return localStorage.getItem('authToken');
}

export function setAuthToken(token) {
  localStorage.setItem('authToken', token);
}

export function removeAuthToken() {
  localStorage.removeItem('authToken');
}
