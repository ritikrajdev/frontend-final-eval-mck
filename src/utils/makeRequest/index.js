import axios from 'axios';
import {
  ERROR_ROUTE,
  ERROR_WITH_CODE_ROUTE,
  LOGIN_ROUTE,
} from '../../constants/routes';
import { getAuthToken, removeAuthToken } from '../auth';

export async function makeRequest(apiEndpoint, dyamicConfig, navigate) {
  dyamicConfig = dyamicConfig ?? {};

  const authToken = getAuthToken();
  if (authToken) {
    if (dyamicConfig.headers) {
      dyamicConfig.headers['Authorization'] = authToken;
    } else {
      dyamicConfig.headers = { Authorization: authToken };
    }
  }

  try {
    const response = await axios({
      ...dyamicConfig,
      ...apiEndpoint,
    });

    return response.data;
  } catch (err) {
    if (navigate) {
      if (err.response?.status) {
        if (err.response.status === 401) {
          removeAuthToken();
          navigate(LOGIN_ROUTE);
        } else {
          navigate(ERROR_WITH_CODE_ROUTE(err.response.status));
        }
      } else {
        navigate(ERROR_ROUTE);
      }
    } else {
      throw err;
    }
  }
}
