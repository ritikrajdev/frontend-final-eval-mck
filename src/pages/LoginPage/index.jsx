import React, { useEffect } from 'react';
import { useNavigate } from 'react-router-dom';

import circleWhite from '../../assets/images/circle_white.svg';
import uploadIllustration from '../../assets/images/undraw_upload.svg';
import Button from '../../components/elements/Button';
import Input from '../../components/elements/Input';
import { loginApiEndpoint } from '../../constants/apiEndpoints';
import { ERROR_ROUTE, HOME_ROUTE } from '../../constants/routes';
import { getAuthToken, setAuthToken } from '../../utils/auth';
import { makeRequest } from '../../utils/makeRequest';
import './LoginPage.css';

export default function LoginPage() {
  const [error, setError] = React.useState('');
  const navigate = useNavigate();

  useEffect(() => {
    if (getAuthToken()) navigate(HOME_ROUTE);
  }, []);

  async function login(e) {
    e.preventDefault();
    const formData = new FormData(e.target);
    const email = formData.get('email');
    const password = formData.get('password');

    try {
      const { token } = await makeRequest(loginApiEndpoint, {
        data: {
          email,
          password,
        },
      });

      console.log(token);
      setAuthToken(token);

      navigate(HOME_ROUTE);
    } catch (error) {
      console.log(error);
      if (Math.floor(error.response?.status / 100) === 4) {
        setError(error.response.data.message);
      } else {
        navigate(ERROR_ROUTE);
      }
    }
  }

  return (
    <div className='login-page'>
      <div className='app-info'>
        <img
          src={circleWhite}
          style={{
            top: '10px',
            right: '10px',
          }}
        />
        <img
          src={circleWhite}
          style={{
            top: '120px',
            right: '120px',
            width: '100px',
            height: '100px',
          }}
        />
        <h1>
          DESIGN APIs fast,
          <br />
          Manage content easily.
        </h1>

        <img className='img-illustration' src={uploadIllustration} alt='' />

        <img
          src={circleWhite}
          style={{
            bottom: '10px',
            left: '120px',
            width: '100px',
            height: '100px',
          }}
        />
        <img
          src={circleWhite}
          style={{
            bottom: '80px',
            left: '10px',
          }}
        />
      </div>

      <div className='login'>
        <h2>Log in to your account</h2>
        <form onSubmit={login}>
          <Input label='Email' type='email' name='email' />
          <br />
          <Input label='Password' type='password' name='password' />
          <br />
          <p style={{ padding: '0 10px', color: 'red' }}>{error}</p>
          <br />
          <center>
            <Button type='submit'>Log in</Button>
          </center>
        </form>
      </div>
    </div>
  );
}
