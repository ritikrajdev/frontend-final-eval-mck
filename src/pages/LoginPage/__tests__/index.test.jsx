import { fireEvent, render, waitFor } from '@testing-library/react';
import React from 'react';
import { useNavigate } from 'react-router-dom';
import LoginPage from '..';

import { getAuthToken } from '../../../utils/auth';
import { makeRequest } from '../../../utils/makeRequest';

jest.mock('react-router-dom');
jest.mock('../../../utils/auth');
jest.mock('../../../utils/makeRequest');

describe('LoginPage', () => {
  it('renders correctly', () => {
    const navigate = jest.fn();
    useNavigate.mockReturnValue(navigate);

    const view = render(<LoginPage />);
    expect(view.asFragment()).toMatchSnapshot();
  });

  it('navigates to home page if user is already logged in', () => {
    const navigate = jest.fn();
    useNavigate.mockReturnValue(navigate);
    getAuthToken.mockReturnValue('token');

    render(<LoginPage />);
    expect(navigate).toHaveBeenCalledWith('/');
  });

  it('navigates to error page if login fails', async () => {
    const navigate = jest.fn();
    useNavigate.mockReturnValue(navigate);
    makeRequest.mockRejectedValue({
      response: {
        status: 500,
      },
    });

    const view = render(<LoginPage />);
    const emailInput = view.container.querySelector('input[name="email"]');
    const passwordInput = view.container.querySelector(
      'input[name="password"]'
    );
    const submitButton = view.container.querySelector('button[type="submit"]');

    emailInput.value = 'a@b.com';
    passwordInput.value = 'pass';
    fireEvent.click(submitButton);

    await waitFor(() => {
      expect(navigate).toHaveBeenCalledWith('/error');
    });
  });

  it('navigates to home page if login succeeds', async () => {
    const navigate = jest.fn();
    useNavigate.mockReturnValue(navigate);
    makeRequest.mockResolvedValue({
      response: {
        token: 'token',
      },
    });

    const view = render(<LoginPage />);
    const emailInput = view.container.querySelector('input[name="email"]');
    const passwordInput = view.container.querySelector(
      'input[name="password"]'
    );
    const submitButton = view.container.querySelector('button[type="submit"]');
    emailInput.value = 'a@b.com';
    passwordInput.value = 'pass';
    fireEvent.click(submitButton);

    await waitFor(() => {
      expect(navigate).toHaveBeenCalledWith('/');
    });
  });

  it('shows error message if login fails', async () => {
    const navigate = jest.fn();
    useNavigate.mockReturnValue(navigate);
    makeRequest.mockRejectedValue({
      response: {
        status: 400,
        data: {
          message: 'error message',
        },
      },
    });

    const view = render(<LoginPage />);
    const emailInput = view.container.querySelector('input[name="email"]');
    const passwordInput = view.container.querySelector(
      'input[name="password"]'
    );
    const submitButton = view.container.querySelector('button[type="submit"]');
    emailInput.value = 'a@b.com';
    passwordInput.value = 'pass';
    fireEvent.click(submitButton);

    await waitFor(() => {
      expect(view.container).toHaveTextContent('error message');
    });
  });
});
