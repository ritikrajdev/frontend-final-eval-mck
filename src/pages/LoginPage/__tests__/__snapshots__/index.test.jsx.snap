// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage renders correctly 1`] = `
<DocumentFragment>
  <div
    class="login-page"
  >
    <div
      class="app-info"
    >
      <img
        src="circle_white.svg"
        style="top: 10px; right: 10px;"
      />
      <img
        src="circle_white.svg"
        style="top: 120px; right: 120px; width: 100px; height: 100px;"
      />
      <h1>
        DESIGN APIs fast,
        <br />
        Manage content easily.
      </h1>
      <img
        alt=""
        class="img-illustration"
        src="undraw_upload.svg"
      />
      <img
        src="circle_white.svg"
        style="bottom: 10px; left: 120px; width: 100px; height: 100px;"
      />
      <img
        src="circle_white.svg"
        style="bottom: 80px; left: 10px;"
      />
    </div>
    <div
      class="login"
    >
      <h2>
        Log in to your account
      </h2>
      <form>
        <div
          class="custom-input"
        >
          <label>
            Email
          </label>
          <br />
          <input
            name="email"
            type="email"
            value=""
          />
        </div>
        <br />
        <div
          class="custom-input"
        >
          <label>
            Password
          </label>
          <br />
          <input
            name="password"
            type="password"
            value=""
          />
        </div>
        <br />
        <p
          style="padding: 0px 10px; color: red;"
        />
        <br />
        <center>
          <button
            class="custom-button"
            type="submit"
          >
            Log in
          </button>
        </center>
        <br />
        <br />
        <center />
      </form>
    </div>
  </div>
</DocumentFragment>
`;
